lua/wire/stools/expression2.lua

```diff
@@ -259,156 +256,120 @@ if SERVER then
+		toent.DownloadAllowedPlayers[ply] = nil
 
 		local uploadandexit = net.ReadBit() ~= 0
-		local numpackets = net.ReadUInt(16)
 
 		if not IsValid(toent) or toent:GetClass() ~= "gmod_wire_expression2" then
 			WireLib.AddNotify(ply, "Invalid entity specified to wire_expression2_download_wantedfiles. Download aborted.", NOTIFY_ERROR, 7, NOTIFYSOUND_DRIP3)
 			return
 		end
 
-		if not wantedfiles[ply] then wantedfiles[ply] = {} end
-		table.insert(wantedfiles[ply], net.ReadData(net.ReadUInt(32)))
-		if numpackets <= #wantedfiles[ply] then
-			local ok, ret = pcall(WireLib.von.deserialize, table.concat(wantedfiles[ply]))
-			wantedfiles[ply] = nil
+		net.ReadStream(ply, function(data)
+			if not IsValid(toent) then return end
+
+			local ok, ret = pcall(WireLib.von.deserialize, data)
 			if not ok then
 				WireLib.AddNotify(ply, "Expression 2 download failed! Error message:\n" .. ret, NOTIFY_ERROR, 7, NOTIFYSOUND_DRIP3)
 				print("Expression 2 download failed! Error message:\n" .. ret)
 				return
 			end
 
 			WireLib.Expression2Download(ply, toent, ret, uploadandexit)
-		end
+		end)
 	end)
 
 	-- ------------------------------------------------------------
 	-- Serverside Receive
 	-- ------------------------------------------------------------
-	local uploads = WireLib.RegisterPlayerTable()
-	local upload_ents = WireLib.RegisterPlayerTable()
 	net.Receive("wire_expression2_upload", function(len, ply)
-		local toent = Entity(net.ReadUInt(16))
-		local numpackets = net.ReadUInt(16)
+		local toentID = net.ReadUInt(16)
+		local toent = Entity(toentID)
 
 		if (not IsValid(toent) or toent:GetClass() ~= "gmod_wire_expression2") then
-			if uploads[ply] then -- this is to prevent notification spam due to the net library automatically limiting its own transfer rate so that the messages arrive late
-				uploads[ply] = nil
-				upload_ents[ply] = nil
-				WireLib.AddNotify(ply, "Invalid Expression chip specified. Upload aborted.", NOTIFY_ERROR, 7, NOTIFYSOUND_DRIP3)
-			end
+			WireLib.AddNotify(ply, "Invalid Expression chip specified. Upload aborted.", NOTIFY_ERROR, 7, NOTIFYSOUND_DRIP3)
 			return
 		end
 
-		if not WireLib.CanTool(ply, toent, "wire_expression2") then
-			WireLib.AddNotify(ply, "You are not allowed to upload to the target Expression chip. Upload aborted.", NOTIFY_ERROR, 7, NOTIFYSOUND_DRIP3)
+		if toent.AwaitingUpload ~= ply then
+			WireLib.AddNotify(ply, "This Expression chip is not awaiting an upload. Upload aborted.", NOTIFY_ERROR, 7, NOTIFYSOUND_DRIP3)
 			return
 		end
+		toent.AwaitingUpload = nil
+		toent.Uploading = true
 
-		if upload_ents[ply] ~= toent then -- a new upload was started, abort previous
-			uploads[ply] = nil
+		if not WireLib.CanTool(ply, toent, "wire_expression2") then
+			WireLib.AddNotify(ply, "You are not allowed to upload to the target Expression chip. Upload aborted.", NOTIFY_ERROR, 7, NOTIFYSOUND_DRIP3)
+			return
 		end
```

Comment:
This CanTool check should be removed because `if toent.AwaitingUpload ~= ply then` check should be enough.

Reply:
Also because Uploading is being set to true before this early return.